text/0036-async-testbench-functions.md

```diff
@@ -0,0 +1,235 @@
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following `Simulator` methods have their signatures updated:
+
+* `add_process(process)`
+* `add_testbench(process, *, background=False)`
+
+The new optional named argument `background` registers the testbench as a background process when true.
+
+Both methods are updated to accept an async function passed as `process`.
+The async function must accept an argument `sim`, which will be passed a simulator context.
+(Argument name is just convention, will be passed positionally.)
+
+The simulator context have the following methods:
+- `get(expr: Value) -> int`
+- `get(expr: ValueCastable) -> any`
+  - Returns the value of `expr` when awaited.
+    When `expr` is a value-castable, the value will be converted through `.from_bits()`.
+- `set(expr: Value, value: ConstLike)`
+- `set(expr: ValueCastable, value: any)`
+  - Set `expr` to `value` when awaited.
+    When `expr` is a value-castable, the value will be converted through `.const()`.
+- `memory_read(instance: MemoryInstance, address)`
+  - Read the value from `address` in `instance` when awaited.
+- `memory_write(instance: MemoryInstance, address, value, mask=None)`
+  - Write `value` to `address` in `instance` when awaited. If `mask` is given, only the corresponding bits are written.
+- `delay(interval)`
+  - Create a trigger object for advancing simulation by `interval` seconds.
+- `tick(domain="sync", *, context=None)`
+  - Create a trigger object for advancing simulation by one tick of `domain`.
+    When an elaboratable is passed to `context`, `domain` will be resolved from its perspective.
+  - If `domain` is asynchronously reset while this is being awaited, `AsyncReset` is raised.
+- `changed(*signals)`
+  - Create a trigger object for advancing simulation until any signal in `signals` changes.
+- `edge(signal, value)`
+  - Create a trigger object for advancing simulation until `signal` is changed to `value`.
+    `signal` must be a 1-bit signal or a 1-bit slice of a signal.
+- `posedge(signal)`
+- `negedge(signal)`
+  - Aliases for `edge(signal, 1)` and `edge(signal, 0)` respectively.
+  
+  `signal` is changed to `value`. `None` is a wildcard and will trigger on any change.
+- `critical()`
+  - Return a context manager that ensures simulation won't terminate in the middle of the enclosed scope.
+
+A trigger object has the following methods:
+- `__await__()`
+  - Advance simulation and return the value(s) of the trigger(s).
+    - `delay` and `tick` triggers return `True` when they are hit, otherwise `False`.
+    - `changed` and `edge` triggers return the current value of the signals they are monitoring.
+- `__aiter__()`
+  - Return an async generator that repeatedly invokes `__await__()` and yields the returned values.
+- `delay(interval)`
+- `tick(domain="sync", *, context=None)`
+- `changed(*signals)`
+- `edge(signal, value)`
+- `posedge(signal)`
+- `negedge(signal)`
```

Comment:
If I construct something like `await sim.delay(interval).posedge(signal)`, what is the behavior?

Reply:
If you combine a `delay` with another trigger, you get a timeout-like behavior. Means you could do something like:
```python
_, timeout = await sim.posedge(signal).delay(0.01)
assert not timeout, "signal must go positive within 10ms"
```

I thought about this after I pushed the previous change, and that `edge` should probably also return whether it was hit like `tick` and `delay`, and not the actual value like `changed`, since we already know the only value it could be.

Combining two delays naturally means that the shorter one will always trigger first.
If they've got the same interval, it's an undefined race like when two clocks tick at the same time step.
While the semantics are clear enough, I don't see a good use for this, so we could forbid adding a second delay trigger to a trigger object that already has one.

Reply:
I can't see implementing this being easy.

Reply:
At a glance, `BaseSimulation` needs another method to cancel a delay trigger added by `wait_interval()`, pysim (and of course cxxsim) needs to implement it and `PyCoroProcess` needs to call it when it's doing `clear_triggers()`.

Reply:
Actually, I guess you could just track the closest `.delay` trigger there is inside the trigger object since it will always fire before the other ones.